### PR TITLE
Recruitment performance generator and worker

### PR DESCRIPTION
**Pull request by inulty-dfe**
## Context

This PR implements the plumbing required to fetch Recruitment Performance data from Bigquery and store it in the apply database.

### Diagram

image

## Changes proposed in this pull request

 - [x] Only get providers that have not had a report generated already
 - [ ] Throttle / stagger the jobs for generating reports
 - [x] Check Bigquery request limiting
 - [x] Get an estimate of how many queries will be sent to BQ each week

### BigQuery quota 
[BigQuery quotas](https://cloud.google.com/bigquery/quotas)

### Queries per week estimate:

Running the query that is used to fetch the provider for whom we should generate a report, we see there are **780** valid providers.
We will make 1 BigQuery request for each of these Providers and that will return all the data we require.

### Throttle / stagger the jobs for generating reports
We could make these queries in batches of 50, 100, 250 or what ever, but I have not tried to optimise for this in this PR. We should test the performance before deciding how to optimise.

## Guidance to review

 - Are there any more edge cases that should be considered?
 - Are the tests written clearly enough?
 - Do we need a FeatureFlag for this? - These jobs will start being enqueued as soon as we merge this otherwise

## Link to Trello card

[Trello Ticket](https://trello.com/c/b9ucaiYB/1609-performance-report-importer-and-worker)

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/models/publications/national_recruitment_performance_report_generator.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  class NationalRecruitmentPerformanceReportGenerator
+    attr_reader :client,
+                :generation_date,
+                :publication_date,
+                :report_expected_time,
+                :cycle_week
+
+    def initialize(cycle_week:, generation_date: Time.zone.now, publication_date: nil)
+      @cycle_week = cycle_week
+      @generation_date = generation_date
```
Comment (elceebee) [inline app/models/publications/national_recruitment_performance_report_generator.rb:11]: I can see the publication date and generation date are the same for now. I'm worried they will always be the same and we'll just have these two identical fields to confuse people. I know I've said it before, but I would prefer we just have one until we are really clear what the difference is and what we are going to do with them. (I will shut up about it now. This is my last comment on the subject)

**Review by elceebee — APPROVED** [review]
I think the code is fine, thoroughly tested. 
I think it's fine without a feature flag -- we have a feature flag for actually showing users the data, so they won't see it until we are ready, but there's no trouble in starting to populate the data. 
I really think the generation date / publication date are just confusing and likely to continue always being the same, but fine if we want to keep them.

**Review by elceebee — APPROVED** [review]
I've tested this locally and everything is working very nicely. 🚀

**Comment by inulty-dfe** [comment]
[Performance Report Importer and Worker](https://trello.com/c/b9ucaiYB/1609-performance-report-importer-and-worker)
